### ISSUE TITLE
Custom yolo training code

### ISSUE THREAD

**Issue by zjjzeng** [ultralytics/ultralytics#21678]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi, I use ultralytics to train detection models on my own dataset. I want more customization and need to modify some source code of the python package. I have already looked through the [advanced customization tutorial](https://docs.ultralytics.com/usage/engine/). It focus on class override method like DetectionTrainer. However, for my task, I hope to change some code in function and parameter which are not under any class. Could you give me some tips? If anyone can provide a simple example, that will help me a lot. FYI, I may need to change something in ultralytics/data/utils.py, including the value of IMG_FORMATS, some lines in function verify_image_label(), and I hope to add a new custom module under data (this is optional) for connivence. Look forward to any reply, thanks.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zjjzeng, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started; an Ultralytics engineer will follow up soon.

We recommend a visit to the Docs for new users where you can find many Python and CLI usage examples and where many of the most common questions may already be answered:
- Docs: https://docs.ultralytics.com/
- Python: https://docs.ultralytics.com/usage/python/
- CLI: https://docs.ultralytics.com/usage/cli/

If this is a 🐛 Bug Report, please provide a minimum reproducible example to help us debug it:
- MRE guide: https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please provide as much information as possible, including:
- What files/functions you plan to modify (e.g., ultralytics/data/utils.py: IMG_FORMATS, verify_image_label())
- A small snippet or diff of the intended changes
- Your training command, environment details, and logs
- Any sample dataset images demonstrating the need for customization
Also verify you are following our Tips for Best Training Results:
- https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord for real-time chat 🎧: https://discord.com/invite/ultralytics
- Discourse for in-depth discussions: https://community.ultralytics.com/
- Subreddit for community threads: https://www.reddit.com/r/Ultralytics/

## Upgrade

Please upgrade to the latest ultralytics package including all requirements in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help us guide you on customizing data utilities and adding modules, please share:
1. The exact changes you intend (diffs or short code snippets)
2. Your current ultralytics version and environment details
3. A minimal script/command that reproduces your need or any issues
4. A tiny sample dataset or paths illustrating the new formats

Thanks for the detailed context—once you provide the above, we can point you to the cleanest extension points or review a PR suggestion 👍

**Comment by Y-T-G** [comment]
You can clone and install a local copy of Ultralytics using:
https://docs.ultralytics.com/quickstart/#__tabbed_1_3

Then you can modify the local source code and the changes will take effect.

**Comment by zjjzeng** [comment]
> You can clone and install a local copy of Ultralytics using: https://docs.ultralytics.com/quickstart/#__tabbed_1_3
> 
> Then you can modify the local source code and the changes will take effect.

Thank you. I know it could take effect, and I have already tried this way. But I think it only works locally, I prefer the more graceful implementation like DetectionTrainer class in the page https://docs.ultralytics.com/usage/engine.

**Comment by Y-T-G** [comment]
To change the behavior of `IMG_FORMATS` and other such variables, you will have to edit source code by cloning your own copy. There's no override for them. You can perform monkey patching, but that will require tracing all the usage of the variable and patching for each file separately.
